### PR TITLE
Change unreachable pattern ICEs to warnings

### PR DESCRIPTION
**Pull request by canndrew**
Allow code with unreachable `?` and `for` patterns to compile.
Add some tests.

### CODE REVIEW

**Comment by canndrew** [comment]
Ugh, bloody make tidy. Every single time.

**Comment by canndrew** [comment]
So to summarise: An unreachable pattern in a `ForLoopDesugar` is now a warning instead of a `span_bug!`. An unreachable pattern in a `TryDesugar` is nothing, not even a warning (because it should be fine to `?` a `Result<!, T>`). `?` now desguars to a match expression which has `#[allow(unreachable_patterns)]` and `#[allow(unreachable_code)]` on it.

**Comment by eddyb** [comment]
cc @rust-lang/compiler

**Review by nikomatsakis — APPROVED** [review]

File: src/librustc/hir/lowering.rs [hunk 1]
```diff
@@ -1833,8 +1835,12 @@ impl<'a> LoweringContext<'a> {
                 ExprKind::Try(ref sub_expr) => {
                     // to:
                     //
+                    // #[allow(unreachable_patterns)]
```
Comment (nikomatsakis) [inline src/librustc/hir/lowering.rs]: Hmm, so, this `#[allow(unreachable_patterns)]` annotation presumably affects the `<expr>`, right? That seems unfortunate.  i.e., if I do `(match { ... })?`, do we still see warnings from the inner `match`?

**Comment by durka** [comment]
Wait, do expr-level `#[allow]` annotations work? [Last I checked, they have no effect.](https://is.gd/ivYz2N)

File: src/librustc/hir/lowering.rs (hunk 1 above)
Comment (canndrew) [inline src/librustc/hir/lowering.rs]: Good point, I'll need to look into this :/

**Comment by arielb1** [comment]
I do have time.

File: src/librustc/hir/lowering.rs (hunk 1 above)
Comment (arielb1) [inline src/librustc/hir/lowering.rs]: Can't you key things directly on the `MatchSource`?

Comment (canndrew) [inline src/librustc/hir/lowering.rs]: I still seem to get warnings without these. At least the `unreachable_code` attribute seems to have an effect. I was thinking I'll change the code to this:

```rust
match Carrier::translate(<expr>) {
    #[allow(unused)]
    Ok(val) => val,
    #[allow(unused)]
    Err(err) => return Carrier::from_error(From::from(err)),
}
```

Comment (arielb1) [inline src/librustc/hir/lowering.rs]: The `unreachable_code` attribute is indeed fine - no user-defined code exists under it. It's the `unreachable_patterns` attribute that is the problem - could you check why is it needed?

**Comment by nikomatsakis** [comment]
@arielb1 ok :) either way. In any case, I think we both agree that we should try to avoid having warnings get squelched in user code *somehow*. I agree with the suggestion of using the `MatchSource` to squelch the lints, does that not work?

**Comment by nikomatsakis** [comment]
@canndrew I just want to clarify something. Given https://github.com/rust-lang/rust/pull/39151, is this PR necessary for backwards compatibility? I'm trying to assess the priority of landing it with respect to the release timeline.

**Comment by arielb1** [comment]
@nikomatsakis

The PR is required. #39151 just gates the uninhabitableness of references, it does not revert the changes for matching `Result<X, Void>`.

File: src/librustc/hir/lowering.rs [hunk 2]
```diff
@@ -1849,14 +1855,34 @@ impl<'a> LoweringContext<'a> {
...
+                                attr::mk_spanned_list_item(e.span, allow_ident, vec![uc_spanned])
+                            };
+                            attr::mk_spanned_attr_outer(e.span, attr::mk_attr_id(), allow)
+                        };
+                        let attrs = From::from(vec![val_attr]);
+                        let val_expr = P(self.expr_ident_with_attrs(e.span,
+                                                                    val_ident,
+                                                                    val_pat.id,
+                                                                    attrs));
+                        let val_block = P(self.block_expr(val_expr));
```
Comment (arielb1) [inline src/librustc/hir/lowering.rs]: Do we need the `block_expr`/`expr_block` pair?

Comment (canndrew) [inline src/librustc/hir/lowering.rs]: No.. for some reason I thought we did.

**Comment by canndrew** [comment]
Sorry for spamming travis.

I've now changed the try desugaring to this:

```rust
match Carrier::translate(<expr>) {
    Ok(val) => #[allow(unreachable_code)] val,
    Err(err) => #[allow(unreachable_code)] return Carrier::from_error(From::from(err)),
}
```

That seems to be all we need to crush the warnings.

File: src/librustc/hir/lowering.rs [hunk 3]
```diff
@@ -1833,8 +1834,12 @@ impl<'a> LoweringContext<'a> {
...
                     // to:
                     //
+                    // #[allow(unreachable_patterns)]
                     // match Carrier::translate(<expr>) {
-                    //     Ok(val) => val,
+                    //     Ok(val) => {
+                    //         #[allow(unreachable_code)]
+                    //         val
+                    //     }
                     //     Err(err) => return Carrier::from_error(From::from(err))
```
Comment (arielb1) [inline src/librustc/hir/lowering.rs]: nit: update the comment here

File: src/librustc/hir/lowering.rs [hunk 4]
```diff
@@ -1849,11 +1854,29 @@ impl<'a> LoweringContext<'a> {
...
+
+                    // Ok(val) => { #[allow(unreachable_code)] val }
                     let ok_arm = {
                         let val_ident = self.str_to_ident("val");
                         let val_pat = self.pat_ident(e.span, val_ident);
-                        let val_expr = P(self.expr_ident(e.span, val_ident, val_pat.id));
+                        let val_expr = P(self.expr_ident_with_attrs(e.span,
+                                                                    val_ident,
+                                                                    val_pat.id,
+                                                                    From::from(attrs.clone())));
```
Comment (arielb1) [inline src/librustc/hir/lowering.rs]: I prefer `ThinVec::from`, but whatever you want.

File: src/librustc_const_eval/check_match.rs [hunk 5]
```diff
@@ -329,9 +322,7 @@ fn check_arms<'a, 'tcx>(cx: &mut MatchCheckCtxt<'a, 'tcx>,
                                                             hir_pat.id, diagnostic);
                         },
 
-                        hir::MatchSource::TryDesugar => {
-                            span_bug!(pat.span, "unreachable try pattern")
-                        },
+                        hir::MatchSource::TryDesugar => {}
```
Comment (arielb1) [inline src/librustc_const_eval/check_match.rs]: A comment would be nice, for example:
```Rust
// unreachable patterns in a try expression occur when one of the arms
// is of an uninhabited type, which is OK.
```

**Review by arielb1 — CHANGES_REQUESTED** [review]
nice modulo comment nits.

**Pull request merged**